### ISSUE TITLE
prefer-destructuring and reassignment of variables

### ISSUE THREAD

**Issue by zloirock** [eslint/eslint#7886]
For example, we have code like:
```js
let error;

if (data.detail) {
    if (data.detail.error) {
        error = data.detail.error;
    } else {
        doSomething();
        error = data.detail;
    }
} else {
    error = data.error;
}

doSomethingWith(error);
```
`prefer-destructuring` rule reports 2 issues for this code. I have no ideas how to beautifully refactor this code for following this rule, only renaming a variable and it does not look like a good solution.

So I propose to add an option for reporting about it only for variable declarations, not for assignment expressions.

**Comment by fasttime** [comment]
Maybe the documentation for this new rule could use some more examples, but in this case, all it really takes to apply destructuring are parentheses:

```js
let error;

if (data.detail) {
    if (data.detail.error) {
        ({ error } = data.detail);
    } else {
        doSomething();
        error = data.detail;
    }
} else {
    ({ error } = data);
}

doSomethingWith(error);
```

**Comment by zloirock** [comment]
@fasttime yep, thanks, I know about parentheses, but it looks too ugly - original code looks clearer for me =)

**Comment by not-an-aardvark** [comment]
Seems like a good idea to me. :+1:

Maybe we could have two separate options (one for variable declarations and one for assignment expressions):

```yml
prefer-destructuring: [error, {VariableDeclarator: true, AssignmentExpression: true}] # current behavior
prefer-destructuring: [error, {VariableDeclarator: true, AssignmentExpression: false}] # only check declarations
# ... etc
```

**Comment by vitorbal** [comment]
Agreed, I think that's a good idea

**Comment by nunofgs** [comment]
I also had something along the lines of:

```js
function foo(err) {
  if (foobar) {
    err = foobar.err; // false positive on `prefer-destructuring` here
  }

  return err;
}
```

I had to disable the rule because of this :(

Is anyone working on the `AssignmentExpression` feature?

**Comment by vitorbal** [comment]
@nunofgs I don't think anyone is currently working on this. If you are willing to put in the leg-work for a PR, we can muster some support to get this accepted!

**Comment by not-an-aardvark** [comment]
I'll champion this proposal.

One subtlety that I forgot to mention in https://github.com/eslint/eslint/issues/7886#issuecomment-271177755: The `prefer-destructuring` rule already has options for arrays and objects, and we would probably want to make those independent of the variable and assignment options. Maybe something like this would be a better schema:

```yml
prefer-destructuring: [error, {
  VariableDeclarator: [{ arrays: true, objects: true }],
  AssignmentExpression: [{ arrays: true, objects: false }] # don't require object destructuring for assignments
}, {
  enforceForRenamedProperties: true
}]
```

**Comment by thealjey** [comment]
@not-an-aardvark and what about function arguments and import/export statements? I think those can also benefit from this rule a lot

**Comment by not-an-aardvark** [comment]
@thealjey The rule currently does not support function parameters and import/export statements. It might be worth changing that, but it seems like that is orthogonal to this discussion. Feel free to create a new issue for that.

**Comment by nunofgs** [comment]
@not-an-aardvark any news on this? 😇

**Comment by not-an-aardvark** [comment]
I haven't started this yet. I'm planning to work on this eventually when I have time, but if someone else wants to work on implementing it before then, feel free.

**Comment by VictorHom** [comment]
Hey @not-an-aardvark I can take a look at this if you haven't started it.

**Comment by not-an-aardvark** [comment]
Feel free!
